single_kernel_mongo/events/backups.py

```diff
@@ -0,0 +1,252 @@
+            )
+            success_action_with_info_log(
+                logger,
+                event,
+                action,
+                {"backup-status": f"backup started. backup id: {backup_id}"},
+            )
+        except Exception as e:
+            fail_action_with_error_log(
+                logger,
+                event,
+                action,
+                str(e),
+            )
+
+    def _on_list_backups_action(self, event: ActionEvent):
+        action = "list-backups"
+        check, reason = self.pass_sanity_checks()
+        if not check:
+            fail_action_with_error_log(logger, event, action, reason)
+            return
+
+        try:
+            self.manager.assert_can_list_backup()
+        except InvalidPBMStatusError as e:
+            fail_action_with_error_log(logger, event, action, str(e))
+            return
+
+        try:
+            formatted_list = self.manager.list_backup_action()
+            success_action_with_info_log(logger, event, action, {"backups": formatted_list})
+        except ListBackupError as e:
+            fail_action_with_error_log(logger, event, action, str(e))
+            return
+
+    def _on_restore_action(self, event: ActionEvent):
+        action = "restore"
+
+        backup_id = str(event.params.get("backup-id", ""))
+        remapping_pattern = str(event.params.get("remap-pattern", ""))
+
+        if self.dependent.state.upgrade_in_progress:
+            fail_action_with_error_log(
+                logger, event, action, "Restoring a backup is not supported during an upgrade."
+            )
+            return
+        check, message = self.pass_sanity_checks()
+        if not check:
+            fail_action_with_error_log(logger, event, action, message)
+            return
+        if not self.charm.unit.is_leader():
+            fail_action_with_error_log(
+                logger, event, action, "The action can be run only on a leader unit."
+            )
+            return
+
+        try:
+            self.manager.assert_can_restore(
+                backup_id,
+                remapping_pattern,
+            )
+        except (InvalidPBMStatusError, InvalidArgumentForActionError) as e:
+            fail_action_with_error_log(logger, event, action, str(e))
+            return
+
+        try:
+            self.manager.restore_backup(backup_id=backup_id, remapping_pattern=remapping_pattern)
+            self.charm.status_manager.set_and_share_status(
+                MaintenanceStatus(f"restore started/running, backup id:'{backup_id}'")
+            )
+            success_action_with_info_log(
+                logger, event, action, {"restore-status": "restore started"}
+            )
+        except ResyncError:
+            raise
+        except RestoreError as restore_error:
+            fail_action_with_error_log(logger, event, action, str(restore_error))
+
+    def pass_sanity_checks(self) -> tuple[bool, str]:
```

Comment:
There is: Protocols :)
I'm however weighing if this should live in the event handler or in the manager, and for now my heart is really balanced. If I had to make a decision for that, I would have it in the managers because it makes sense that the manager is able to take decisions while the event handler should mainly be a dispatch/decision patter.

Reply:
> I'm however weighing if this should live in the event handler or in the manager, and for now my heart is really balanced. If I had to make a decision for that, I would have it in the managers because it makes sense that the manager is able to take decisions while the event handler should mainly be a dispatch/decision pattern.

If its possible to add a protocol that would be great! I think it would be best in the event handler, but also down for it in managers for the same reasoning you outlined :) 

> For the healthy property, I'm as well weighing this vs a get_status method that returns a status value but this is something I want to bring along the status ticket

Lets postpone the healthy for now

Reply:
I'll add it in the `feat/status-handler` branch on which I'll be working today.
This leaves me some time to clean the interface and reduces the load on this PR.
I think the best is in the managers because the managers are the one which are knowledgeable of the current status of the manager while the event handlers should `only` dispatch events and react to the errors raised.